src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DotnetDataCollectionLauncher.cs

```diff
@@ -0,0 +1,193 @@
+            // If already running with the dotnet executable, use it; otherwise pick up the dotnet available on path.
+            // Wrap the paths with quotes in case dotnet executable is installed on a path with whitespace.
+            if (currentProcessPath.EndsWith(DotnetProcessNameXPlat, StringComparison.OrdinalIgnoreCase)
+                || currentProcessPath.EndsWith(DotnetProcessName, StringComparison.OrdinalIgnoreCase))
+            {
+                currentProcessFileName = currentProcessPath;
+            }
+            else
+            {
+                currentProcessFileName = this.GetDotnetHostFullPath();
+            }
+
+            if (EqtTrace.IsVerboseEnabled)
+            {
+                EqtTrace.Verbose("DotnetDataCollectionLauncher: Full path of dotnet.exe is {0}", currentProcessFileName);
+            }
+
+            var dataCollectorAssemblyPath = Path.Combine(currentWorkingDirectory, DataCollectorProcessName);
+            commandLineArguments.Insert(0, dataCollectorAssemblyPath);
+
+            dataCollectorFileName = Path.GetFileNameWithoutExtension(dataCollectorAssemblyPath);
+
+            // .NET core host manager is not a shared host. It will expect a single test source to be provided.
+            var args = "exec";
+
+            // Probe for runtimeconfig and deps file for the test source
+            var runtimeConfigPath = Path.Combine(currentWorkingDirectory, string.Concat(dataCollectorFileName, ".runtimeconfig.json"));
+
+            if (this.fileHelper.Exists(runtimeConfigPath))
+            {
+                var argsToAdd = " --runtimeconfig \"" + runtimeConfigPath + "\"";
+                args += argsToAdd;
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("DotnetDataCollectionLauncher: Adding {0} in args", argsToAdd);
+                }
+            }
+            else
+            {
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("DotnetDataCollectionLauncher: File {0}, doesnot exist", runtimeConfigPath);
+                }
+            }
+
+            // Use the deps.json for test source
+            var depsFilePath = Path.Combine(currentWorkingDirectory, string.Concat(dataCollectorFileName, ".deps.json"));
+            if (this.fileHelper.Exists(depsFilePath))
+            {
+                var argsToAdd = " --depsfile \"" + depsFilePath + "\"";
+                args += argsToAdd;
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("DotnetDataCollectionLauncher: Adding {0} in args", argsToAdd);
+                }
+            }
+            else
+            {
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("DotnetDataCollectionLauncher: File {0}, doesnot exist", depsFilePath);
+                }
+            }
+
+            commandLineArguments.Insert(0, args);
```

Comment:
IList.Insert is O(n). We're doing it multiple times in this logic. Why not use String.Concat and String.Join(commandLineArguments)?